### PR TITLE
Move special ops into interpreter

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33265 Clean up isinstance flags
* **#32889 Move special ops into interpreter**
* #32804 Remove prim::Constant op
* #32791 interpreter handling for varargs to remove need for looking at Node

Common primitive ops that have special inputs make it very hard to
serialize the bytecode for mobile because information about how the
op behaves is hidden in the Node*. This changes how we handle the following
ops so that they are encoded as their own interpreter bytecodes.

```
    USES NODE: prim::TupleUnpack(...) -> (...)
    USES NODE: prim::TupleSlice(...) -> (...)
    USES NODE: prim::TupleConstruct(...) -> (...)
    USES NODE: prim::ListUnpack(...) -> (...)
    USES NODE: prim::ListConstruct(...) -> (...)
    USES NODE: prim::DictConstruct(...) -> (...)
    USES NODE: prim::Constant() -> (...)
    USES NODE: prim::isinstance(...) -> (...)
    USES NODE: prim::CreateObject(...) -> (...)
    USES NODE: prim::fork(...) -> (...)
    USES NODE: aten::warn(str message, *, int stacklevel=2) -> () # need stack level information, so ideally in interpreter so it can look at the stack
```

This leaves a state where the _only_ remaining Node*-consuming builtins
are things that are only introduced during JIT optimization and will
not appear in mobile code.

Serialization of bytecode can now be made to directly write the CodeImpl
object without modification.

Differential Revision: [D19673157](https://our.internmc.facebook.com/intern/diff/D19673157)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit ff65dfed:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ff65dfed3c8f8dfd53a5113dde01f0023e4f46d5).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4394641) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93940698) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93940698))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
Removing tools/autograd/templates/python_variable_methods_dispatch.h 
Removing tools/autograd/templates/python_torch_functions_dispatch.h 
Removing tools/autograd/templates/python_nn_functions_dispatch.h 
Failed to merge submodule third_party/fbgemm (not checked out) 
Auto-merging third_party/fbgemm 
CONFLICT (submodule): Merge conflict in third_party/fbgemm 
Auto-merging test/test_jit_py3.py 
Auto-merging test/test_jit.py 
Auto-merging test/backward_compatibility/check_backward_compatibility.py 
CONFLICT (content): Merge conflict in test/backward_compatibility/check_backward_compatibility.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/csrc/jit/vararg_functions.cpp [hunk 1]
```diff
@@ -2,33 +2,16 @@
...
-  vals.reserve(num_inputs);
-  for (size_t i = stack_size - num_inputs; i < stack_size; ++i) {
-    vals.emplace_back(std::move(stack[i]).toTensor());
-  }
-  drop(stack, num_inputs);
-  push(stack, std::move(vals));
-}
 
-void tupleUnpackFunc(int num_outputs, Stack& stack) {
+void tupleUnpack(Stack& stack) {
```
Comment (bhosmer) [inline torch/csrc/jit/vararg_functions.cpp]: Is the story here that `num_outputs` was never really necessary? AFAICT the logic is the same, mod the `AT_ERROR` below.

**Review by bhosmer — APPROVED** [review]
LGTM, accepting to unblock but just re-request if the tuple/list types idea winds up yielding new code you'd like to have looked over.

File: torch/csrc/jit/vararg_functions.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/jit/vararg_functions.cpp]: correct

**Comment by zdevito** [comment]
re-request because I made changes to mobile/interpreter.cpp to get the tests to pass again.

**Review by bhosmer — APPROVED** [review]
Mobile changes LGTM

**Comment by apaszke** [comment]
What does a "Node*-consuming builtin" means in this case? That the operation that needs to be executed depends on the node attributes?

**Pull request closed without merging**